### ISSUE TITLE
Rule Request: [disable_rationale]

### ISSUE THREAD

**Issue by acecilia** [realm/SwiftLint#2655]
### New Issue Checklist

- [ *] Updated SwiftLint to the latest version
- [* ] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

Please describe the rule idea, format
this issue's title as `Rule Request: [Rule Name]` and describe:

1. Why should this rule be added? Share links to existing discussion about what
   the community thinks about this.

I would like to have an explanation of why a rule is disabled next to it

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

* would:

```
// swiftlint:disable missing_docs
```

* would not:
```
// This rule is disabled because of blablabla
// swiftlint:disable missing_docs
```

3. Should the rule be configurable, if so what parameters should be configurable?
No configuration required.

4. Should the rule be opt-in or enabled by default? Why?
   See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a rule as opt-in.
Opt-in.

### Custom rule implementation
At the same time I am proposing this rule, I tried to implement it as a custom rule. I did not managed, and I would appreciate some help. My main idea is:

(1) Find a line containing 'swiftlint:disable' ==> (2) Check that the previous line is a comment

I thought I could do this with a lookbehind regex. The closest I got was the following regex:

```
custom_rules:
  disabled_rules_rationale:
    regex: "(?<!\/\/[^\n]*$\n)^[^\n]*swiftlint:disable"
    match_kinds:
      - comment
    message: "All swiftlint disable comments should include an explanation on top"
```

In [https://regexr.com](regexr) the outcome is the desired one:

![screenshot 2019-02-21 at 16 25 08](https://user-images.githubusercontent.com/7938671/53180118-49a01b80-35f5-11e9-9b25-652b6cf12701.png)

But using the same regex in swiftlint does not result the same (note there are no errors):
![screenshot 2019-02-21 at 16 26 38](https://user-images.githubusercontent.com/7938671/53180244-7f450480-35f5-11e9-94f6-ae817347cd96.png)

Some final notes:
* I am using `[^\n]` instead of `.` in the regex because I did not manage to make `.` work in swiftlint.
* The `[^\n]*` part of the lookbehind group in the `"(?<!\/\/[^\n]*$\n)^[^\n]*swiftlint:disable"` regex is the one making things not work: using `"(?<!\/\/$\n)^[^\n]*swiftlint:disable"` has the following result, which is very close to the intended behaviour:

![screenshot 2019-02-21 at 16 30 40](https://user-images.githubusercontent.com/7938671/53180551-0db98600-35f6-11e9-830d-c7c76657b8a9.png)

Thanks!

**Comment by acecilia** [comment]
After looking at the `vertical_whitespace_closing_braces` rule regex, I think I managed to implement a working solution:

```
  disable_comments_should_have_rationale:
    regex: "^(?![^\n]*\/\/)([^\n]*\n[^\n]*swiftlint:disable[^\n]*\n)(?!\/\/)"
    message: "When disabling a rule, the rationale should be added as a comment in the previous or next line"
```

This allows to attach the rationale to the previous or next line, so you do not fall into problems when using `:next`, `:this` or ':previous`.

**Comment by Dschee** [comment]
@acecilia Thank you for suggesting this rule. I think it would make sense for some teams to use this, while I don't think it should be enabled by default. My personal take on disabling SwiftLint rules is that it should only done if the reason is obvious (like for false positives) – otherwise one should just live with the warnings until somebody has the time to fix them. But again, I can see this become useful in some contexts.

One thing I didn't see in your examples is what we should do when rules are disabled in a code line, like here:

```swift
func someFunction() { // swiftlint:disable:this function_body_length
    // some code here
    
    var someVar = 5
    var x = 1 // swiftlint:disable:this identifier_name
    var y = 2 // swiftlint:disable:this identifier_name
    var someOtherVar = 7

    // some code here
}
```

**Comment by Dschee** [comment]
Also another problem I see with this is when multiple rules are disabled in a single line. How can the check ensure that explanations are given for all disablings? Or consider this: When one rule was disabled and a reason was provided for that, then I add another disabling to the same line, how will this rule remind me to update the rationale as well?

My current thinking is like this: Most people won't need this rule. But those who will might also want to make sure every disabling has a rationale to it. So maybe we should think of a structure like this:

```swift
// identifier_name: false positive
// line_length: a URL should be kept together for easier copy & paste
// swiftlint:disable:this identifier_name line_length
```

Allowing to combine both reasons into one line:

```swift
// identifier_name: false positive / line_length: a URL should be kept together for easier copy & paste
// swiftlint:disable:this identifier_name line_length
```

We should also consider making sure that the disablings and rationales stay consistent, so when a disabling is removed, we should warn against keeping the rationale.

**Comment by Dschee** [comment]
Also I believe it would be useful to have some predefined groups which can be used by a shorthand. The most common groups I can see coming are `**f**alse positive` (should be ovious), `externally **m**anaged` (for example when the code is generated by a tool), `**d**eprecated` (code that will be removed anyways), `needs **r**efactoring` (old code part that was written without SwiftLint that should be refactored anyways) and `**e**xception` (when applying the rule in this case would make the code worse). 

We could think up a list such as the one suggested above and give them shorthands and only require a rationale for the `exception` type. The result could look something like this:

```swift
// line_length: a URL should be kept together for easier copy & paste
// swiftlint:disable:this identifier_name(f) line_length(e)
```

Note that `f` stands for `false positive` and `e` for `exception` in this example.

**Comment by acecilia** [comment]
@Dschee thanks for the feedback! Some ideas I got while reading:

- Adding a rationale per rule is great but: how do you know when a comment is a leftover rationale VS a normal comment? For example, for the `disable` and `enable` comments the keyword `swiftlint` is appended at the start.
- The shortlands you proposed looks good to me, but there are some problems I can think of:
  * Implementing them probably requires reimplementing the detection and processing of the disable comments. I will need to test it, or maybe you can confirm this.
  * Which common rationales should be implemented? 
    * I agree with the `false positive` and the `deprecated`.  
    * I do not think the `externally managed` makes sense: if you do not want some files to be linted because they are autogenerated, you should better add them to the `excluded` section of the configuration file. 
    * Also, the `needs refactoring` seems to me like an easy workaround to avoid truly refactoring the code: it is more like `somebody will need to refactor this code, but it is not going to be me`. This is exactly what I am trying to avoid with this rule: encourage the developer to think about when is truly necessary to disable a rule, and not just disable it because he is in a rush. I am afraid you can easily end up with multiple `disable - needs refactoring` all around the app which will never be refactor.

The scenario where I see this rule being useful is as follows: when having a team of developers, you usually want to keep some level of style and code quality through the codebase. For such a thing, you setup swiftlint in a CI. Also, because you do not want your lint warnings accumulating over time, you turn the severity of all the rules to `error`, and lint only the files appearing on the diff of the pull requests (so you are not forced to a massive refactor of legacy code, but better lint the code over time). At this point is very easy for everybody in the team to work around the linter by using the disable commands all over the place. Having a rationale helps the code reviewers assess when a `disable` comment makes sense, and encourages the developers to think a bit further about when a `disable` comment should be used.

And one last comment. Overall, all the feedback ideas you proposed are improvements over the initial idea, but I am not sure how much time I am going to be able to put into fulfilling them. Would it be possible to agree on some basic functionality to get it merge, and iterate on it? I would say that all your comments are additive, so we could have the actual PR merged (just detecting rationale comments on top or bottom of the disable comments), and then work from there.

**Comment by jpsim** [comment]
I support doing something along these lines, with the following thoughts:

1. Seems like this would be more of a top-level SwiftLint configuration than a _rule_, since it would apply to all rules.
2. I don't see any reason to use regular expressions, since we already know where the comment blocks are, we could look to see if the disable command is accompanied by other comments.
